prover/src/server.rs

```diff
@@ -0,0 +1,231 @@
+                serde_json::from_reader(body_bytes);
+
+            if let Err(err) = json_req {
+                let payload = serde_json::to_vec(&JsonRpcResponseError {
+                    jsonrpc: "2.0".to_string(),
+                    id: 0.into(),
+                    error: JsonRpcError {
+                        // parser error
+                        code: -32700,
+                        message: err.to_string(),
+                    },
+                })
+                .unwrap();
+                let mut resp = Response::new(Body::from(payload));
+                set_headers(resp.headers_mut(), false);
+                return Ok(resp);
+            }
+
+            let json_req = json_req.unwrap();
+            let result: Result<serde_json::Value, String> =
+                handle_method(json_req.method.as_str(), &json_req.params, &shared_state).await;
+            let payload = match result {
+                Err(err) => {
+                    serde_json::to_vec(&JsonRpcResponseError {
+                        jsonrpc: "2.0".to_string(),
+                        id: json_req.id,
+                        error: JsonRpcError {
+                            // internal server error
+                            code: -32000,
+                            message: err,
+                        },
+                    })
+                }
+                Ok(val) => serde_json::to_vec(&JsonRpcResponse {
+                    jsonrpc: "2.0".to_string(),
+                    id: json_req.id,
+                    result: Some(val),
+                }),
+            };
+            let mut resp = Response::new(Body::from(payload.unwrap()));
+            set_headers(resp.headers_mut(), false);
+            Ok(resp)
+        }
+
+        // serve CORS headers
+        (&Method::OPTIONS, "/") => {
+            let mut resp = Response::default();
+            set_headers(resp.headers_mut(), true);
+            Ok(resp)
+        }
+
+        // everything else
+        _ => {
+            let mut not_found = Response::default();
+            *not_found.status_mut() = StatusCode::NOT_FOUND;
+            Ok(not_found)
+        }
+    }
+}
+
+async fn handle_method(
+    method: &str,
+    params: &[serde_json::Value],
+    shared_state: &SharedState,
+) -> Result<serde_json::Value, String> {
+    match method {
+        // enqueues a task for computating proof for any given block
+        "proof" => {
+            let options = params.get(0).ok_or("expected struct ProofRequestOptions")?;
+            let options: ProofRequestOptions =
+                serde_json::from_value(options.to_owned()).map_err(|e| e.to_string())?;
+
+            shared_state
+                .get_or_enqueue(&options)
+                .await
+                .map(|result| serde_json::to_value(result?).map_err(|e| e.to_string()))
+                .unwrap_or_else(|| Ok(serde_json::Value::Null))
+        }
+        // TODO: Add the abilitity to abort the current task.
```

Comment:
I file an issue for that. I haven't yet looked for a way to interrupt a busy thread - especially one inside the tokyo worker pool without breaking some stuff. Because the prover code will eventually run long inside a synchronous context and I don't know yet if interrupting threads like this will lead to memory leaks or similar

Reply:
Will try to investigate on that!